### PR TITLE
GH-92123: Move _elementtree heap types to module state

### PR DESCRIPTION
**Pull request by erlend-aasland**
* Issue: gh-92123

### CODE REVIEW

File: Modules/_elementtree.c [hunk 1]
```diff
@@ -2140,7 +2155,8 @@ elementiter_next(ElementIterObject *it)
                 continue;
             }
 
-            assert(Element_Check(extra->children[child_index]));
+            elementtreestate *st = ET_STATE_GLOBAL;
+            assert(Element_Check(st, extra->children[child_index]));
```
Comment (kumaraditya303) [inline Modules/_elementtree.c:2160]:
```suggestion
#ifndef NDEBUG
            elementtreestate *st = ET_STATE_GLOBAL;
            assert(Element_Check(st, extra->children[child_index]));
#endif
```

**Review by kumaraditya303 — APPROVED** [review]

File: Modules/_elementtree.c [hunk 2]
```diff
@@ -69,13 +69,6 @@ static void _clear_joined_ptr(PyObject **p)
     }
 }
 
-/* Types defined by this extension */
-static PyTypeObject *Element_Type;
-static PyTypeObject *ElementIter_Type;
-static PyTypeObject *TreeBuilder_Type;
-static PyTypeObject *XMLParser_Type;
```
Comment (arhadthedev) [inline Modules/_elementtree.c:76]: Probably, these variables should be deleted from the c-analyzer list as well:

https://github.com/python/cpython/blob/78d1676336491ffc0279da0a9f6d7e91252be783/Tools/c-analyzer/cpython/globals-to-fix.tsv#L428-L431

Comment (erlend-aasland) [reply Modules/_elementtree.c:76]: Very good catch; thanks.

**Comment by erlend-aasland** [comment]
Thanks for the reviews, Kumar and Oleg!

For the path forward, there are three options: either pass state as a parameter, store it in the type structs. I've created draft PRs for the two alternatives. Of course, a third path is a combination of both.

**Comment by arhadthedev** [comment]
> For the path forward, there are three options: either pass state as a parameter, store it in the type structs.

Both variants ultimately call `PyModule_GetState`, right? [with *(1)* caching its result into a field of `self` and *(2)* plainly calling it each time on demand]

If so, we need an informed decision by benchmarking some `_elementtree` method in both options. It should show whether direct structure access of *(1)* is faster than preheated call to `PyModule_GetState` of *(2)*.

**Comment by erlend-aasland** [comment]
> Both variants ultimately call `PyModule_GetState`, right?

`PyModule_GetState` is called for module level functions, `PyType_GetModuleState` for class methods, and `PyType_GetModuleByDef` plus `PyModule_GetState` for slots. The two former are inexpensive lookups, the latter is slightly slower. Caching state in `self` results in more memory usage (might only be important for the Element type). I'd be surprised if it is possible to measure the performance impact of this, but sure let's do some benchmarking.

**Comment by erlend-aasland** [comment]
I ran the pyperformance `xml_etree` benchmark towards all three alternatives (1: pass state around as a parameter, 2: store state in the heap type contexts, 3: store state in TreeBuilder and XMLParser contexts, but pass it around for Element and its iter). I could not measure a significant difference any of the alternatives.

**Comment by arhadthedev** [comment]
In this case I'm absolutely neutral towards all variants and can't suggest any preference.

**Comment by kumaraditya303** [comment]
It won't matter as expected as most of the time as long as there isn't much stack spilling the performance should be identical.

**Pull request merged**